src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs

```diff
@@ -327,12 +327,28 @@ private bool IsCommitCharacter(char ch)
             var selectedItem = GetExternallyUsableCompletionItem(model.SelectedItem);
             if (selectedItem.IsBuilder)
             {
-                return char.IsLetterOrDigit(ch);
+                return char.IsLetterOrDigit(typedChar);
             }
 
-            var filterText = GetCurrentFilterText(model, selectedItem);
+            var snapshot = this.TextView.TextSnapshot;
+            var editSpan = model.GetCurrentSpanInSnapshot(model.GetSubjectBufferFilterSpanInViewBuffer(selectedItem.FilterSpan), snapshot);
+            var spanStart = (int)editSpan.Start;
+            var caretPosition = (int)GetCaretPointInViewBuffer();
+
+            var filterText = snapshot.GetText(Span.FromBounds(spanStart, caretPosition));
+            var isCommitCharacter = GetCompletionRules().IsCommitCharacter(selectedItem, typedChar, filterText);
+
+            // When a character, even appears to be a commit character, was typed at the begining of the filter span when it was not empty,
+            // it's most likely that the user just realized he/she forgot to do something, pressed Ctrl+Left and typed it.
+            // i.e. blah| -> |blah -> !|blah
+            // We should not commit the selected item in the above case.
+            // But be careful we still want completion in the following case: A a = new | -> A a = new (|
+            if (isCommitCharacter && caretPosition == spanStart + 1 && editSpan.Length > 1)
```

Comment:
I'd like @jasonmalinowski to weigh in on this.  Normally we map things to the subject buffer and make our determinations there.  But in this code we map things up to the view buffer.  This seems very different from our normal way of handling positions, views, buffers, and positions.   I'd prefer we stick with how we generally do things unless there is a strong reason to use the view's buffer.  I'm curious what jason thinks.